### PR TITLE
Autogenerate contract types

### PR DESCRIPTION
**Pull request by CoreyMartin**
hardhat is able to autogenerate contract types so that we can reference contract methods and arguments with type safety and autocomplete in tests. For example if a test provides the wrong argument type to a contract method running `npm run test` will now make that clear instead of failing for a less obvious reason. This PR follows hardhat instructions here to set it up https://hardhat.org/guides/typescript.html#type-safe-smart-contract-interactions

VS Code before:

Screen Shot 2022-05-09 at 4 22 14 PM

VS Code after:

Screen Shot 2022-05-09 at 4 22 27 PM

### CODE REVIEW

File: hardhat.config.ts [hunk 1]
```diff
@@ -38,4 +42,12 @@ module.exports = {
...
         },
     },
+    typechain: {
+        outDir: "types",
+        target: "ethers-v5",
+        // should overloads with full signatures like deposit(uint256) be generated always, even if there are no overloads?
+        alwaysGenerateOverloads: false,
+        // optional array of glob patterns with external artifacts to process (for example external libs from node_modules)
+        externalArtifacts: ["externalArtifacts/*.json"],
+    },
```
Comment (CoreyMartin) [inline hardhat.config.ts:52]: Just taking the defaults from here https://github.com/dethcrypto/TypeChain/tree/master/packages/hardhat

File: .gitignore [hunk 2]
```diff
@@ -50,4 +50,6 @@ tenderly.yaml
 # Github #
 ##########
 
-secrets/
\ No newline at end of file
+secrets/
+
+types/
```
Comment (CoreyMartin) [inline .gitignore:55]: These are meant to be temporary files generated for the tests so we are not committing them

File: package.json [hunk 3]
```diff
@@ -7,13 +7,14 @@
...
-        "lint:js": "eslint . --ext .ts",
-        "lint:js:fix": "eslint --fix .",
+        "lint": "npm run lint:ts && npm run lint:sol",
+        "lint:fix": "npm run lint:ts:fix && npm run lint:sol:fix",
+        "lint:ts": "eslint . --ext .ts",
+        "lint:ts:fix": "eslint --fix .",
         "lint:sol": "solhint contracts/**/*.sol && prettier --check \"contracts/**/*.sol\"",
-        "lint:sol:fix": "solhint contracts/**/*.sol --fix && prettier --write \"contracts/**/*.sol\""
+        "lint:sol:fix": "solhint contracts/**/*.sol --fix && prettier --write \"contracts/**/*.sol\"",
+        "test": "npm run typechain && hardhat test",
```
Comment (CoreyMartin) [inline package.json:16]: `hardhat test` alone will generate the types but for some reason the tests are running too fast and not seeing the types. Fixed by running the commands separately

File: test/ContractTestContext.ts [hunk 4]
```diff
@@ -12,85 +24,96 @@ async function getContractFactories (factoryNames: string[]): Promise<{ [K: stri
...
+    leverageEngine: LeverageEngine;
+    leverageAllocator: LeverageAllocator;
+    positionToken: PositionToken;
+    parameterStore: ParameterStore;
     // External contracts
-    externalOUSD;
+    externalOUSD: Contract;
+}
+
+export async function buildContractTestContext (): Promise<ContractTestContext> {
```
Comment (CoreyMartin) [inline test/ContractTestContext.ts:47]: It's a bit of a pain getting the class types to work nicely when an async method like setup is required. Instead just converting to an async function that returns the context as an object.

**Review by yotamdaniel — APPROVED** [review]
Great improvement to our workflow, thank you!

**Pull request merged**